### ISSUE TITLE
Use deriving strategies for Hapistrano's Monad

### ISSUE THREAD

**Issue by ibarrae** [stackbuilders/hapistrano#180]
As it is now, we Hapistrano is using the following monad stack:
https://github.com/stackbuilders/hapistrano/blob/f3fbe9a34cde5c645cbb4d51a48647a80ea00b92/src/System/Hapistrano/Types.hs#L49

Using [deriving strategies](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/deriving_strategies.html) The monad could be implemented like the following:
``` haskell
newtype Hapistrano (a :: Type) = Hapistrano {unHapistrano :: Config -> IO a}
  deriving (Functor, Applicaitve, Monad, MonadThrow) via (ReaderT Config IO)
```

That could remove the current monad stack and also lift operations between them, exceptions could be caught at [top level](https://github.com/stackbuilders/hapistrano/blob/2bfde28d698eef3b31a1b5d5e1938e271b5735ea/src/System/Hapistrano.hs#L54) during the execution of commands and could simplify the code a bit.

For backward compatibility, we might use CPP as well to keep the monad stack or choose the new monad as well.

Let me know how does this sound :smile:

**Comment by DavidMazarro** [comment]
This looks doable. I have a question regarding the usage of `MonadCatch` here (which would need to be added to the instances of the newtype): to preserve the ability to handle the current error we have, i.e. the pair `(Failure, Maybe Release)`, we would define something like this:
```haskell
newtype FailureException = FailureException { unFailureEx :: (Failure, Maybe Release) }
  deriving (Show, Typeable)

instance Exception FailureException
```
is that right? In that case, I think the refactor could be quite easy to carry out.

**Comment by ibarrae** [comment]
Yeah, I think that would do. You could also use `anyclass` as well for the instance of `Exception`.

``` haskell
newtype FailureException = FailureException { unFailureEx :: (Failure, Maybe Release) }
  deriving (Show, Typeable)
  deriving anyclass Exception
```

**Comment by CristhianMotoche** [comment]
Hey @ibarrae Would you mind if I pick this one? I haven't touch this new type of deriving and I'd love to learn more about it.

**Comment by ibarrae** [comment]
For sure @CristhianMotoche :rocket:

**Comment by CristhianMotoche** [comment]
Solved in #204
